### ISSUE TITLE
user not recognized when in cmd

### ISSUE THREAD

**Issue by gitramro** [sequelize/sequelize-auto#360]
I'm trying to type this on cmd (Windows 10):

sequelize-auto -o "./models" -d ArgentinaDB -h XX.XXX.XXX.XX -u admincontractsapi -p 62010 -x SuperPass20* -e mssql

And im getting this back

{ SequelizeConnectionError: Login failed for user ''.
    at Connection.<anonymous> (C:\Users\ROMEROR\AppData\Roaming\npm\node_modules\sequelize-auto\node_modules\sequelize\lib\dialects\mssql\connection-manager.js:109:16)
    at Connection.emit (events.js:189:13)
    at Connection.message (C:\Users\ROMEROR\AppData\Roaming\npm\node_modules\tedious\lib\connection.js:1972:18)
    at Connection.dispatchEvent (C:\Users\ROMEROR\AppData\Roaming\npm\node_modules\tedious\lib\connection.js:1172:36)
    at MessageIO.messageIo.on (C:\Users\ROMEROR\AppData\Roaming\npm\node_modules\tedious\lib\connection.js:1045:14)
    at MessageIO.emit (events.js:189:13)
    at Message.message.on (C:\Users\ROMEROR\AppData\Roaming\npm\node_modules\tedious\lib\message-io.js:32:14)
    at Message.emit (events.js:194:15)
    at endReadableNT (C:\Users\ROMEROR\AppData\Roaming\npm\node_modules\tedious\node_modules\readable-stream\lib\_stream_readable.js:1077:12)
    at process._tickCallback (internal/process/next_tick.js:63:19)
  name: 'SequelizeConnectionError',
  message: 'Login failed for user \'\'.',
  parent:
   { ConnectionError: Login failed for user ''.
       at ConnectionError (C:\Users\ROMEROR\AppData\Roaming\npm\node_modules\tedious\lib\errors.js:13:12)
       at Parser.tokenStreamParser.on.token (C:\Users\ROMEROR\AppData\Roaming\npm\node_modules\tedious\lib\connection.js:751:23)
       at Parser.emit (events.js:189:13)
       at Parser.parser.on.token (C:\Users\ROMEROR\AppData\Roaming\npm\node_modules\tedious\lib\token\token-stream-parser.js:27:14)
       at Parser.emit (events.js:189:13)
       at addChunk (C:\Users\ROMEROR\AppData\Roaming\npm\node_modules\tedious\node_modules\readable-stream\lib\_stream_readable.js:297:12)
       at readableAddChunk (C:\Users\ROMEROR\AppData\Roaming\npm\node_modules\tedious\node_modules\readable-stream\lib\_stream_readable.js:279:11)
       at Parser.Readable.push (C:\Users\ROMEROR\AppData\Roaming\npm\node_modules\tedious\node_modules\readable-stream\lib\_stream_readable.js:240:10)
       at Parser.Transform.push (C:\Users\ROMEROR\AppData\Roaming\npm\node_modules\tedious\node_modules\readable-stream\lib\_stream_transform.js:139:32)
       at doneParsing (C:\Users\ROMEROR\AppData\Roaming\npm\node_modules\tedious\lib\token\stream-parser.js:80:14)
       at token (C:\Users\ROMEROR\AppData\Roaming\npm\node_modules\tedious\lib\token\infoerror-token-parser.js:48:5)
       at call.lineNumber (C:\Users\ROMEROR\AppData\Roaming\npm\node_modules\tedious\lib\token\infoerror-token-parser.js:13:19)
       at awaitData (C:\Users\ROMEROR\AppData\Roaming\npm\node_modules\tedious\lib\token\stream-parser.js:179:7)
       at Parser.awaitData (C:\Users\ROMEROR\AppData\Roaming\npm\node_modules\tedious\lib\token\stream-parser.js:103:7)
       at Parser.readUInt32LE (C:\Users\ROMEROR\AppData\Roaming\npm\node_modules\tedious\lib\token\stream-parser.js:176:10)
       at parser.readBVarChar.procName (C:\Users\ROMEROR\AppData\Roaming\npm\node_modules\tedious\lib\token\infoerror-token-parser.js:12:90) message: 'Login failed for user \'\'.', code: 'ELOGIN' },
  original:
   { ConnectionError: Login failed for user ''.
       at ConnectionError (C:\Users\ROMEROR\AppData\Roaming\npm\node_modules\tedious\lib\errors.js:13:12)
       at Parser.tokenStreamParser.on.token (C:\Users\ROMEROR\AppData\Roaming\npm\node_modules\tedious\lib\connection.js:751:23)
       at Parser.emit (events.js:189:13)
       at Parser.parser.on.token (C:\Users\ROMEROR\AppData\Roaming\npm\node_modules\tedious\lib\token\token-stream-parser.js:27:14)
       at Parser.emit (events.js:189:13)
       at addChunk (C:\Users\ROMEROR\AppData\Roaming\npm\node_modules\tedious\node_modules\readable-stream\lib\_stream_readable.js:297:12)
       at readableAddChunk (C:\Users\ROMEROR\AppData\Roaming\npm\node_modules\tedious\node_modules\readable-stream\lib\_stream_readable.js:279:11)
       at Parser.Readable.push (C:\Users\ROMEROR\AppData\Roaming\npm\node_modules\tedious\node_modules\readable-stream\lib\_stream_readable.js:240:10)
       at Parser.Transform.push (C:\Users\ROMEROR\AppData\Roaming\npm\node_modules\tedious\node_modules\readable-stream\lib\_stream_transform.js:139:32)
       at doneParsing (C:\Users\ROMEROR\AppData\Roaming\npm\node_modules\tedious\lib\token\stream-parser.js:80:14)
       at token (C:\Users\ROMEROR\AppData\Roaming\npm\node_modules\tedious\lib\token\infoerror-token-parser.js:48:5)
       at call.lineNumber (C:\Users\ROMEROR\AppData\Roaming\npm\node_modules\tedious\lib\token\infoerror-token-parser.js:13:19)
       at awaitData (C:\Users\ROMEROR\AppData\Roaming\npm\node_modules\tedious\lib\token\stream-parser.js:179:7)
       at Parser.awaitData (C:\Users\ROMEROR\AppData\Roaming\npm\node_modules\tedious\lib\token\stream-parser.js:103:7)
       at Parser.readUInt32LE (C:\Users\ROMEROR\AppData\Roaming\npm\node_modules\tedious\lib\token\stream-parser.js:176:10)
       at parser.readBVarChar.procName (C:\Users\ROMEROR\AppData\Roaming\npm\node_modules\tedious\lib\token\infoerror-token-parser.js:12:90) message: 'Login failed for user \'\'.', code: 'ELOGIN' } }
Done!

I already tried using quotes, double quotes, etc

Also using this (read it from the docs):

var SequelizeAuto = require('sequelize-auto')

var auto = new SequelizeAuto('ArgentinaDB', 'admincontractsapi', 'SuperPass20*', {
    host: 'XX.XXX.XXX.XX',
    dialect: 'mssql',
    directory: 'models',
    port: 62010,
    tables: ['Inflow Table']
})

auto.run(function (err) {
    if (err) throw err;
  
    console.log(auto.tables); // table list
    console.log(auto.foreignKeys); // foreign key list
  });

im getting the same issue, btw with this same credentials Im already in on sequelize and its working fine.

**Comment by darrensapalo** [comment]
I was trying to do exactly the same thing (MSSQL too) and stumbled on this problem yesterday. It's an old piece of unmaintained code. The problem is that the default authentication is not being configured. 

This is what I did to make it work.

### Resolution

1. Open up the `connection.js` of the tedious.js library, which handles the authentication process for sequelize. You can find out which file it is by checking your stacktrace. For your case, it is:

```
C:\Users\ROMEROR\AppData\Roaming\npm\node_modules\tedious\lib\connection.js
```

2. Look for the line that checks whether `config.authentication` is undefined and insert the following authentication object:

```
    let authentication;
    // Insert the following authentication object
    config.authentication = {
      type: "default"
    };
    if (config.authentication !== undefined) {
      if (typeof config.authentication !== 'object' || config.authentication === null) {
        throw new TypeError('The "config.authentication" property must be of type Object.');
      }
```

3. There's a part there that checks for the type of the authentication, whether `azure-active-directory-password`, `azure-active-directory-access-token`, or `default`. We just set it to default, but now the problem is, it isn't loading the username and password you passed in the CLI. Look for the following code snippet; there should be two of occurrences of these definitions of the `authentication` object:

```
        authentication = {
          type: 'default',
          options: {
            userName: options.userName,
            password: options.password
          }
        };
```

4. Replace `options.userName` with `config.userName`, and likewise for the password.

```
        authentication = {
          type: 'default',
          options: {
            userName: config.userName,
            password: config.password
          }
        };
```

Run your command in CLI and it should generate the sequelize models.

**Note:**

For safety, uninstall your global `tedious` installation since you tinkered with the code. You might encounter unexpected problems if you forget you modified the source code for `tedious`, so might as well uninstall it after generating your MS SQL models.

**Comment by SebastianPfliegel** [comment]
Thanks @darrensapalo! I got this error yesterday.

Is this something that should be fixed permanently on tedious or sequelize-auto side?

**Comment by darrensapalo** [comment]
@SebastianPfliegel I didn't take long to inspect the cause, but judging based on the activity of the two npm projects, I'd wager that it's most likely `sequelize-auto` having problems. Most probably misconfiguring its access of tedious? Or incorrectly pinning the tedious version its version supports?

I find it unlikely that [tedious](https://www.npmjs.com/package/tedious) which is designed for accessing MSSQL is the one having authentication problems in their code. Their builds also pass all of their tests, which I covers authentication.

In any case, it appears that the author of sequelize-auto is on leave, but there are plenty of pull requests for this project.

**Comment by ATLTed** [comment]
Updating Sequlize and fixing the lodash references fixed this problem for me.

**Comment by CoderDream** [comment]
> lodash references

Could you give me some details, thanks.

**Comment by cmincarelli** [comment]
I will try to describe how I dealt with this. Following ATLTed's lead.. I forked the repo, then cloned locally, updated the packages and fixed the lodash references.

UPDATE packages: use yarn (not npm): "yarn upgrade-interactive --latest". update all the old modules.

REPLACE all references of [helpers.]Sequelize.Utils._ with require("lodash")

`var _ = Sequelize.Utils._;` will become  `var _ = require('lodash');`

Then I ran the utility in the ./bin directory:

`./bin/sequelize-auto -o "./models" -d blah blah blah`

**Comment by CLiNTPELiX** [comment]
Hi!
Despite your solution, I can't connect also...
Can you make send/gist/attach me connection.js modified by your own.

Thanks a lot!

**Comment by codecranium** [comment]
Had the same issue.

**Comment by geilsonrm** [comment]
tedious@5.0.3 saved here!

**Comment by spiffai** [comment]
I'm having this problem and am unable to get this to work using your fix @darrensapalo  - could you send a copy of your  changes to me? I've tried everything I can think of and this just isn't working. I am able to connect to my database using the exact same information via sequelize, just not sequelize-auto (says user is '').

**Comment by darrensapalo** [comment]
@spiffai Hey there. You can check out the modifications I made for the connection.js under tedious [here](https://gist.github.com/darrensapalo/93ca5e913c00441a514ad24d5fccecdf).

Particularly, this `connection.js` file was found in this part of my machine:

`/Users/my-username-for-my-machine/.npm-packages/lib/node_modules/tedious/lib/connection.js`

It may vary based on your environment/installation. 

Hope that helps buddy.

**Comment by bburns** [comment]
I was getting the same error so wound up using **sequelizer** instead - https://github.com/andyforever/sequelizer. It might use sequelizer-auto under the hood but it works on Windows connecting to a MSSQL db.

**Comment by LiamKarlMitchell** [comment]
Still a problem on latest.

Node v14.3.0
Microsoft SQL Server 2016 (SP2-GDR) (KB4532097) - 13.0.5102.14 (X64)   Dec 31 2019 22:39:35   Copyright (c) Microsoft Corporation  Express Edition (64-bit) on Windows 10 Pro 10.0 <X64> (Build 19041: ) (Hypervisor) 

    "sequelize": "^6.0.0-beta.6",
    "tedious": "^8.3.0"

```bash
sequelize-auto -o "./models" -d db -h 127.0.0.1 -U sa -p 1433 -x pass -e mssql
```

**Comment by steveschmitt** [comment]
I think this is fixed in `sequelize-auto@0.5.0`  
Please try it and close this issue if so.  Thanks!

**Comment by steveschmitt** [comment]
Problem was that `sequelize-auto` was using an old version of `sequelize` that was incompatible with current `tedious`

Fixed in `sequelize-auto@0.5.1`
